custom_components/localtuya/__init__.py

```diff
@@ -174,63 +100,60 @@ def _device_discovered(device):
         product_key = device["productKey"]
 
         # If device is not in cache, check if a config entry exists
+        entry = async_config_entry_by_device_id(hass, device_id)
+        if entry is None:
+            return
+
         if device_id not in device_cache:
-            entry = async_config_entry_by_device_id(hass, device_id)
-            if entry:
+            if entry and device_id in entry.data[CONF_DEVICES]:
                 # Save address from config entry in cache to trigger
                 # potential update below
-                device_cache[device_id] = entry.data[CONF_HOST]
+                host_ip = entry.data[CONF_DEVICES][device_id][CONF_HOST]
+                device_cache[device_id] = host_ip
 
         if device_id not in device_cache:
             return
 
-        entry = async_config_entry_by_device_id(hass, device_id)
-        if entry is None:
-            return
+        dev_entry = entry.data[CONF_DEVICES][device_id]
 
-        updates = {}
+        new_data = entry.data.copy()
+        updated = False
 
         if device_cache[device_id] != device_ip:
-            updates[CONF_HOST] = device_ip
+            updated = True
+            new_data[CONF_DEVICES][device_id][CONF_HOST] = device_ip
             device_cache[device_id] = device_ip
 
-        if entry.data.get(CONF_PRODUCT_KEY) != product_key:
-            updates[CONF_PRODUCT_KEY] = product_key
+        if dev_entry.get(CONF_PRODUCT_KEY) != product_key:
+            updated = True
+            new_data[CONF_DEVICES][device_id][CONF_PRODUCT_KEY] = product_key
 
         # Update settings if something changed, otherwise try to connect. Updating
         # settings triggers a reload of the config entry, which tears down the device
         # so no need to connect in that case.
-        if updates:
-            _LOGGER.debug("Update keys for device %s: %s", device_id, updates)
-            hass.config_entries.async_update_entry(
-                entry, data={**entry.data, **updates}
+        if updated:
+            _LOGGER.debug(
+                "Updating keys for device %s: %s %s", device_id, device_ip, product_key
             )
-        elif entry.entry_id in hass.data[DOMAIN]:
-            _LOGGER.debug("Device %s found with IP %s", device_id, device_ip)
-
-            device = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE]
-            device.async_connect()
+            new_data[ATTR_UPDATED_AT] = str(int(time.time() * 1000))
+            hass.config_entries.async_update_entry(entry, data=new_data)
+            device = hass.data[DOMAIN][TUYA_DEVICES][device_id]
```

Comment:
Also for some reason; *after* I had done migration and successful cloud setup; this appears. Possibly not initalised to hass.data[DOMAIN][TUYA_DEVICES] = {} or similar at some point by an async method?
Something else?
Unclear.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.9/asyncio/selector_events.py", line 1029, in _read_ready
    self._protocol.datagram_received(data, addr)
  File "/config/custom_components/localtuya/discovery.py", line 70, in datagram_received
    self.device_found(decoded)
  File "/config/custom_components/localtuya/discovery.py", line 79, in device_found
    self._callback(device)
  File "/config/custom_components/localtuya/__init__.py", line 140, in _device_discovered
    device = hass.data[DOMAIN][TUYA_DEVICES][device_id]
KeyError: 'tuya_devices'
```

Reply:
OK I believe we do have a race condition here... does it persist after you restart HA?

Reply:
Solved